### ISSUE TITLE
Hiding partial content collections, and collection variables

### ISSUE THREAD

**Issue by vitorio** [grow/grow#216]
Hi!

In #82 we talked about partial content collections, with empty blueprints and no paths or views.  These work!  But today I was working on auto-generating an index of contents, iterating over all the collections and then all the documents in each, and that partial content collection blew it up.

I was expecting a partial content collection (empty blueprint, no front matters, no paths, no views) to also not show up in g.collections(), but it did.  I currently have to treat it like a reserved word and skip it.  I'm not sure if this was a reasonable expectation on my part or not.

Also, it would be nice if:
- Collections could have orders, like $order in documents, and that `g.collections()` could take an `order_by` like `g.docs` can
- Collections could provide the directory name of the collection (not the title), so I don't have to do `{% set dc = doc.collection|string %}{% set dname = dc.split('"')[1] %}` to pass something appropriate to `g.docs`
- Collections could have a $titles array, like documents can, and/or arbitrary key:value pairs

If any of these seem like good ideas, I can try taking a stab at a pull request for them.

**Comment by vitorio** [comment]
Okay, revisiting this a bit more, it looks like a partial content collection with an empty or missing `_blueprint.yaml` and documents that all start with underscores don't show up in `g.collections()`.  (The content collection itself doesn't need to lead with an underscore, just the documents.)  So that part's sorted for me.  The other stuff would be nice, though!

**Comment by jeremydw** [comment]
These suggestions sound great -- I think `g.collections` should support `$order` in `_blueprint.yaml` and should also support the same `$titles` behavior that docs do. That'd be great.

For the directory name, yes, I agree on this too. I think for consistency with `Document` objects and, well Unix conventions (https://en.wikipedia.org/wiki/Basename), we should name this `basename`. So one could do `{{doc.collection.basename}}`. Thoughts on that?

Long story short, all sounds good, and I'm happy to take these on or pass them to you. Just let me know. Thanks!

**Comment by jeremydw** [comment]
A roundup of changes since this issue was filed:

- Implemented `Collection.basename` to the directory name of the collection.
- Implemented `Collection.titles()`, similar to `Document.titles()`. The syntax for this is:

```
$title: Default Title
$titles:
  nav: Nav Title

{{doc.collection.title}}  # Default Title
{{doc.collection.titles()}}  # Default Title
{{doc.collection.titles('nav')}}  # Nav Title
{{doc.collection.ttiles('foo')}} # Default Title
```
- Implemented support for user-defined fields, such as:

```
$path: /pages/{base}/
$view: /views/base.html
foo: bar

{{doc.collection.foo}} # Bar
```

- Built-in fields (`$titles`, `$title`, `$path`, `$view`, etc.) now require a `$` prefix, just like documents. This is backwards-compatible and doesn't require the builtin prefix for old fields, but for all new fields going forward we require `$`.
- Implemented `Collection.order` with a default to `sys.maxint`. This means that you can specify `$order: 1` or `$order: 0` and then sort the collection using Jinja's `sort` filter and unordered documents will go to the end of the list.

Note that I think we should leave `g.collections` unordered by default, and require the use of Jinja's `sort` filter to sort the result. I think this is the right approach since the `sort` filter is more explicit, and also the nature of sorting is often context-specific, so pairing the sort action with a filter seems more correct to me. I think we'll suggest moving `g.docs` to a similar model. Anyway, here's how collection order is used:

```
# /content/col-a/_blueprint.yaml
$order: 2

# /content/col-b/_blueprint.yaml
$order: 1

# /content/col-c/_blueprint.yaml
# no order

{% set cols = g.collections() %}

{{cols}}  # col-a, col-b, col-c (unordered)
{{cols|sort(attribute='order')}}  # col-b, col-a, col-c
{{cols|sort(attribute='order', reverse=true)}} # col-c, col-a, col-b
```

Some extra notes: the reason why we originally designed `g.docs` to support `order_by` was because we thought of this method more like a query, at first. That said, after many months of using this, in reality these methods are more like listing objects rather than querying/searching them, so I think omitting the `order_by` from the list method and instead relying on a filter for sorting in-context makes most sense.

Anyway, I'm going to go ahead and close out this issue based on all of the additions above, but please feel free to re-open it if you think some features are missing and something new needs to be implemented.
